[1 - 3단계 방탈출 사용자 예약] 라젤(문선영) 미션 제출합니다.

RaZel713 opened the pull request:
안녕하세요 ~ 오리 리뷰어님! 😊
이번 방탈출 사용자 예약 미션 리뷰이인 🍋라젤 입니다!

저번 장기 미션에서 리뷰어님께 피드백을 받았었는데, 
레벨 2) 2번째 미션에서도 만나게 되어 너무 기쁘고 반갑습니다 🥰

부족한 부분이 많지만, 잘 부탁드립니다 👍

---

✅ 이번 단계 미션의 변경 사항 커밋 목록

이번 미션 요구사항에 따른 변경 사항은 아래 커밋 범위에 포함되어 있습니다:
👉 [변경 커밋 보기](https://github.com/woowacourse/spring-roomescape-member/pull/169/files/e80d70c4ca32cb1c6ff8c88644f08ade96454f24..2bfa9a6345df9dcec68e8437d9c29ec796649da9)

초기 설정 커밋: e80d70c4ca32cb1c6ff8c88644f08ade96454f24
마지막 커밋: 2bfa9a6345df9dcec68e8437d9c29ec796649da9

※ 이전 미션 코드가 함께 포함되어 있어 전체 diff가 많을 수 있습니다.
위 링크 내 커밋들이 이번 미션에서 수정한 실제 범위입니다.

---

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
convention을 정하고, 책임에 대한 부분을 고민하면서 진행했던 것 같습니다

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [x] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jhan0121 committed:
init: 프로젝트 초기 설정

jhan0121 committed:
Merge branch 'woowacourse:main' into step1

RaZel713 committed:
docs: API 명세 작성

RaZel713 committed:
refactor: Admin 컨트롤러에서 정적 콘텐츠 관리

RaZel713 committed:
feat: theme 테이블 추가 및 관계 설정

RaZel713 committed:
feat: 테마 관리 페이지 응답 메서드 추가

RaZel713 committed:
feat: 테마 도메인 추가

RaZel713 committed:
feat: 테마 DAO 추가

RaZel713 committed:
feat: 테마 조회 기능 추가

jhan0121 committed:
refactor: DTO 패키지 controller로 이동

jhan0121 committed:
feat: ThemeDAO 저장 기능 추가

jhan0121 committed:
feat: 테마 추가 기능 추가

RaZel713 committed:
refactor: 컨트롤러 DTO 패키지 분리

- 요청과 응답 패키지로 분리

RaZel713 committed:
refactor: 컨트롤러 클래스 네이밍 변경 및 패키지 분리

RaZel713 committed:
refactor: 요청 DTO의 역할에 맞는 네이밍 변경

RaZel713 committed:
refactor: CRUD 순서로 메서드 정렬

RaZel713 committed:
refactor: 컨트롤러의 요청 DTO와 서비스 처리 DTO 분리

RaZel713 committed:
test: e2e 테스트 삭제

RaZel713 committed:
feat: 테마 삭제 기능 추가

jhan0121 committed:
fix: URL 경로명 오타 수정

jhan0121 committed:
refactor: 스키마 변경

jhan0121 committed:
feat: 초기 데이터 삽입 sql 추가

jhan0121 committed:
feat: 예약 시, 테마 설정 기능 추가

RaZel713 committed:
feat: 커스텀 예외 처리 기능 추가

RaZel713 committed:
fix: 쿼리 오류 수정

RaZel713 committed:
feat: 커스텀 예외 적용

RaZel713 committed:
feat: 예약, 예약 시간 도메인 생성자에 빈값 검증 추가

RaZel713 committed:
refactor: 미사용 메서드 삭제

RaZel713 committed:
feat: 데이터 추가, 삭제 시 적합한 http 상태 코드 반환

- 데이터 추가 성공 시, 201 CREATED 반환
- 데이터 삭제 성공 시, 204 NO CONTENT 반환
- 데이터 삭제 실패 시, 404 NOT FOUND 반환

RaZel713 committed:
refactor: 예외 처리 로직을 controller에서 service로 이동

RaZel713 committed:
feat: 중복되는 테마 저장 검증 기능 추가

RaZel713 committed:
feat: 예약 시 시간 검증 기능 추가

- 과거 시점 예약 시 예외 반환

jhan0121 committed:
feat: 사용 중인 데이터를 삭제 시도 시 예외 처리 기능 추가

jhan0121 committed:
refactor: controller에 @ResponseStatus 적용

jhan0121 committed:
refactor: controller 네이밍 변경

jhan0121 committed:
test: 임시 데이터 값 수정

jhan0121 committed:
feat: 사용자 예약 페이지 응답 기능 추가

jhan0121 committed:
feat: 예약 가능 시간 조회 기능 추가

RaZel713 committed:
test: 임시 데이터 값 추가

RaZel713 committed:
refactor: 메서드 명 오타 수정

RaZel713 committed:
feat: 인기 테마 목록 조회 기능 추가

RaZel713 committed:
refactor: 코드 순서 정렬

RaZel713 committed:
refactor: 코드 컨벤션 수정

- 클래스 선언부와 필드 선언부 사이에 공백 라인 추가

RaZel713 committed:
test: http 요청 형식 수정

RaZel713 committed:
refactor: final 키워드 적용

RaZel713 committed:
refactor: 테마 도메인 네이밍 변경

jhan0121 committed:
refactor: 커스텀 예외 오버라이드 메서드 삭제

jhan0121 committed:
test: 임시 데이터 값 수정

jhan0121 committed:
test: 기존 테스트 수정

RaZel713 committed:
test: 예약 서비스 기능 테스트 추가

RaZel713 committed:
test: 예약 시간 서비스 기능 테스트 추가

jhan0121 committed:
docs: readme 작성

jhan0121 committed:
refactor: 테마 도메인 네이밍 변경

jhan0121 committed:
refactor: 클래스명 오타 수정

jhan0121 committed:
test: 테마 서비스 기능 테스트 추가

jhan0121 committed:
refactor: 클래스명 오타 수정

jhan0121 committed:
test: 테마 서비스 기능 테스트 추가

jhan0121 committed:
refactor: 서비스 리턴 타입 수정

jhan0121 committed:
refactor: 불필요한 todo 주석 삭제

jhan0121 committed:
Merge remote-tracking branch 'origin/step1' into step1

RaZel713 committed:
refactor: 메서드 네이밍 변경

RaZel713 committed:
refactor: 불필요한 메서드 제거

RaZel713 committed:
refactor: 서비스 메서드 분리

- findById를 생성해 추상화 적용

RaZel713 committed:
test: 미사용 테스트 삭제

RaZel713 committed:
test: 테마 랭킹 조회 http 추가

jinyoungchoi95 commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,36 @@
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+    public static final String ERROR_PREFIX = "[ERROR] ";
```
```suggestion
    private static final String ERROR_PREFIX = "[ERROR] ";
```
상수는 접근을 막아야할 것 같네요.

RaZel713 replied:
상수의 접근 제한자를 private으로 변경하였습니다!

jinyoungchoi95 commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,36 @@
+
+    @ExceptionHandler
+    public ResponseEntity<String> handlePharmaceuticalViolationException(PharmaceuticalViolationException e) {
+        return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).body(ERROR_PREFIX + e.getMessage());
```
422에러는 어떤 상황에 사용되도록 정의되어있나요?

RaZel713 replied:
## Convention) status 코드 정의
- **입력 값 누락 / 빈 값 / null** → 400 Bad Request `InvalidInputException`
- **존재하지 않는 값** → 404 Not Found `NotExistedValueException`
- **이미 존재하는 값을 중복 입력** → 409 Conflict `ExistedDuplicateValueException`
- **사용 중인 상태를 삭제** → 409 Conflict `InUseException`
- **비즈니스 규칙 위반(예: 과거 날짜)** → 422 Unprocessable Content `BusinessRuleViolationException`


### 사용 중인 상태를 삭제 : `InUseException`
프로그래밍 과정에 있어 409 Conflict로 반영했어야 했는데, 422로 잘못 반영되어 있어 
커스텀 예외 (InUseException) 를 추가하고, 409로 반영하였습니다!


### 비즈니스 규칙 위반 : `BusinessRuleViolationException`
`PharmaceuticalViolationException`의 네이밍을 보다 직관적인 `BusinessRuleViolationException`로 변경하였습니다.


> [422 상태 코드 문서](https://developer.mozilla.org/ko/docs/Web/HTTP/Reference/Status/422) 중
> 422 Unprocessable Content
> 하이퍼텍스트 전송 프로토콜(HTTP) 422 Unprocessable Content 응답 상태 코드는 서버가 요청 엔티티의 컨텐츠 형식을 이해했고 요청 엔티티의 문법도 올바르지만 요청된 지시를 처리할 수 없음을 나타냅니다.

위의 문서를 바탕으로 
422 Unprocessable Content는 요청은 이해되었으나 비즈니스 규칙상 허용되지 않는 경우에 적합하다고 판단하여
비즈니스 규칙 위반 시 사용되는 상태 코드로 정의하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/domain/RoomTheme.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public RoomTheme(final String name, final String description, final String thumbnail) {
+        this(0, name, description, thumbnail);
```
0은 의미있는 상수로 변경할 수 있겠네요.

RaZel713 replied:
`private static final int NON_SAVED_STATUS = 0;`

데이터 베이스에 저장되지 않는 상태를 나타내는 의미를 지닌 상수로 변경하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/domain/ReservationTime.java:
```diff
@@ -27,4 +32,18 @@ public long getId() {
     }
+
+    @Override
+    public boolean equals(final Object o) {
```
equals hashCode가 재정의된 곳과 되지 않은 곳의 차이는 무엇인가요?

RaZel713 replied:
``` java
private boolean isBooked(final List<ReservationTime> bookedTimes, final ReservationTime reservationTime) {
        if (bookedTimes.isEmpty()) {
            return false;
        }

        return bookedTimes.contains(reservationTime);
}
```
`bookedTimes.contains(reservationTime)` 코드에서 
`contains`가 `ReservationTime` 객체를 비교할 때, `equals`와 `hashCode`가 재정의되지 않으면, 
기본적으로 참조 주소를 비교하기 때문에 객체의 값을 기준으로 비교하기 위해서 `equals`와 `hashCode`를 재정의 하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/dao/JdbcRoomThemeDAO.java:
```diff
@@ -0,0 +1,89 @@
+    @Override
+    public List<RoomTheme> findPopularThemes(LocalDate start, LocalDate end) {
+
+        final String query = """
+                        SELECT
+                            t.id,
+                            t.theme_name,
+                            t.description,
+                            t.thumbnail
+                        FROM theme AS t
+                        JOIN reservation AS r
+                        ON r.theme_id = t.id
+                        WHERE r.date BETWEEN ? AND ?
+                        GROUP BY t.id
+                        ORDER BY COUNT(r.id) DESC
+                        LIMIT 10
```
쿼리에 인기 테마라는 것은 상위 "10개"가 제한된다가 녹여져있는데 쿼리에 비즈니스 요구사항이 녹아져있으면 무슨 문제가 발생할까요?

RaZel713 replied:
쿼리에 비즈니스 요구사항이 녹아져 있다면,
유연성이 떨어지고, 비즈니스 로직과 데이터 계층이 섞여 테스트와 유지보수가 어려워집니다. 

따라서 해당 요구사항을 서비스 레이어에서 관리하고, 요구사항을 상수 처리하여 요구사항 변경 시,
유연하게 대처할 수 있도록 변경하였습니다.

```java
private static final int POPULAR_THEMES_TOP_COUNT = 10;

public List<RoomTheme> findPopularThemes() {
        LocalDate now = LocalDate.now();
        PopularThemeSelectionCriteria criteria =
                new PopularThemeSelectionCriteria(now, POPULAR_THEME_SELECTION_DURATION);

        List<RoomTheme> popularThemes = themeDAO.findPopularThemes(criteria.getStartDay(), criteria.getEndDay());

        return popularThemes.subList(0, Math.min(POPULAR_THEMES_TOP_COUNT, popularThemes.size()));
}
```

jinyoungchoi95 commented on src/main/java/roomescape/dao/JdbcRoomThemeDAO.java:
dao 테스트가 없는데 dao를 사용하는 클래스는 무엇을 믿고 해당 클래스를 사용할 수 있을까요?

RaZel713 replied:
DAO 테스트를 추가하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/controller/dto/response/AvailableReservationTimeResponse.java:
```diff
@@ -0,0 +1,13 @@
+
+public record AvailableReservationTimeResponse(long id, LocalTime startAt, boolean alreadyBooked) {
+
+    public static AvailableReservationTimeResponse from(ReservationTime reservationTime, boolean alreadyBooked) {
```
영문법상 from은 하나의 파라메터에 대해서 처리하는 것을 의미합니다.

RaZel713 replied:
https://github.com/woowacourse/spring-roomescape-member/pull/169#discussion_r2070119999

``` java
public static AvailableReservationTimeResponse from(AvailableReservationTime availableReservationTime) {
        return new AvailableReservationTimeResponse(
                availableReservationTime.getReservationTime().getId(),
                availableReservationTime.getReservationTime().getStartAt(),
                availableReservationTime.getBookedStatus());
}
```
`AvailableReservationTime` 도메인을 작성하는 과정에 있어 하나의 파라미터로 변경되었습니다.

jinyoungchoi95 commented on src/main/java/roomescape/controller/api/RoomThemeController.java:
`AdminPageController`는 api가 아닌가요? api의 정의가 무엇인가요?

RaZel713 replied:
저는 컨트롤러를 **HTML 페이지를 반환하는 UI 컨트롤러** 와
**데이터를 반환하는 API 컨트롤러** 로 구분하여 관리하는 것이 좋다고 판단하였습니다.

AdminPageController는 넓은 의미에서 API로 해석될 수도 있지만,
사용자에게 UI(HTML 페이지)를 제공하는 목적을 갖고 있으므로
UI 컨트롤러로 보는 것이 타당하다고 생각하였습니다.

---

## API(Application Programming Interface)란
API는 소프트웨어가 서로 데이터를 주고받을 수 있도록 정의된 인터페이스입니다. 
보통은 HTTP를 사용하는 REST API, JSON-RPC, GraphQL API 등 데이터 교환이 목적일 때를 말합니다.

## ❓왜 UI 인가?
`AdminPageController`는 사용자의 요청에 대해 HTML 뷰 템플릿을 반환합니다.
사람(브라우저 사용자) 이 사용하게 되는 UI를 렌더링하기 위한 뷰를 반환하므로, 
“**웹 UI 컨트롤러**”라고 부를 수 있습니다.

## ❓`AdminPageController`는 API가 아닌가?

- 넓은 의미에서 보면 API라고 할 수 있습니다. API는 “어떤 기능을 외부에 제공하는 인터페이스” 라는 개념이므로,
AdminPageController도 “HTML을 제공하는 API” 라고 해석할 수 있습니다.
- 웹 브라우저(Chrome, Edge 등)도 서버의 HTML 응답을 요청하는 “클라이언트”이므로, "HTML을 반환하는 API" 라고 할 수 있습니다.
- 예를 들어, 웹 프레임워크인 Django는 UI용 뷰도 “뷰 API(View API)“라고 부릅니다.

## ❓UI는 API의 하위 개념인가?
• UI는 사람과 시스템 사이의 인터페이스
• API는 시스템과 시스템 사이의 인터페이스
→ 둘 다 “인터페이스”이긴 하지만, 대상이 다르기 때문에 계층적으로 포함되는 관계는 아닙니다.

jinyoungchoi95 replied:
API란 통신에 포함되는 모든 스펙을 의미합니다. json이라고 api다는 아니에요. api의 하위에 json이 있는거죠. view를 반환하는 것도 api의 하위에 있는 것이구요. api의 정의에 대해서 다시한번 고민해보시면 좋겠어요.

`@Controller`라는 것은 모두가 api를 다루는 레이어에 불과합니다. 그 하위에 view resolver와 rest json이 있는 것이에요.

RaZel713 replied:
controller 패키지 명을 rest와 view라고 작성하였습니다!

제가 잘못 생각하고 있었습니다. 알려주셔서 감사합니다.😊

jinyoungchoi95 commented on src/main/java/roomescape/service/dto/ReservationCreation.java:
```diff
@@ -0,0 +1,11 @@
+package roomescape.service.dto;
+
+import java.time.LocalDate;
+import roomescape.controller.dto.request.CreateReservationRequest;
```
import 되었다는 것은 의존한다는 것과 동일한 의미입니다. service 패키지가 controller 패키지에 의존해도되나요?

RaZel713 replied:
계층 간 의존 방향을 컨트롤러 → 서비스 라는 것을 명확히 하여

상위 계층에 대한 역방향 의존을 제거하기 위해

Service에서 도메인을 반환하고 컨트롤러에서 Response DTO로 변환하도록 수정하였습니다.

<details>

<summary>service 패키지가 controller 패키지에 의존하면 안되는 이유</summary>

> 현재의 계층 구조에서는
> Controller → Service → Repository (DAO)와 같은 의존성 방향을 지니고 있습니다.
> 따라서 Service 계층에서 Controller 계층을 의존하는 것은 상위 계층에 대한 역방향 의존으로, 
> 의존성 순환 및 결합도가 증가하게 됩니다.

</details>

jinyoungchoi95 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,52 +1,102 @@
-    public List<Reservation> findAll() {
-        return reservationDAO.findAll();
-    }
+    public ReservationResponse addReservation(final ReservationCreation creation) {
```
controller는 요청 dto에 대해서 `Request`라고 postfix를 달아두는데 여긴 다른 이유가 있나요?

RaZel713 replied:
Controller에서 `Request`, `Response` 라고 postfix를 작성한 이유는
웹 요청 포맷에 맞춘 외부용 DTO임을 나타내기 위하여 `~Request`, `~Response` 라고 작성하였습니다.

그리고 Service에서는 도메인/서비스 로직 중심 내부 요청이므로, 
`Request` 와 `Response`를 사용하지 않고 서비스 계층의 관심사(addReservation)에 맞춰 Creation 이라고 명명하였습니다.

jinyoungchoi95 replied:
외부용 DTO라기에는 `Request`나 `Response`는 외부에 해당하는 것이 아닌 "요청"과 "응답"이라는 용어일뿐인 것 같아요. 즉 해당 용어는 controller에만 국한되어있는 용어가 아닙니다.

RaZel713 replied:
컨트롤러 DTO는 `Create~Request`의 형식으로,
서비스 DTO는 `Create~ServiceRequest`의 형식으로 네이밍 하였습니다!

jinyoungchoi95 commented on src/main/java/roomescape/service/RoomThemeService.java:
```diff
@@ -0,0 +1,73 @@
+                themeCreation.thumbnail());
+        final long id = themeDAO.insert(theme);
+
+        final RoomTheme savedTheme = findById(id);
```
이미 잘 저장했다면 그대로 사용하면 될 것 같은데 조회 쿼리를 또 날릴 필요가 있을까요?

RaZel713 replied:
일관성과 안정성을 고려하여 insert 이후에 반환된 id를 다시 조회해 반환하는 방식을 선택했습니다.

- insert 이후 상태가 DB에 반영되며, 트리거, 디폴트 값, 제약 조건 등으로 인해 DB 내부에서 값이 변경되었을 가능성이 존재합니다.
- 객체 생성 직후와 insert 전후로 다른 레이어나 스레드가 객체를 참조하거나 조작할 수 있는 가능성이 있습니다. 이로 인해 의도치 않게 상태가 변경되거나 오염될 수 있습니다.

jinyoungchoi95 replied:
> 객체 생성 직후와 insert 전후로 다른 레이어나 스레드가 객체를 참조하거나 조작할 수 있는 가능성이 있습니다

말씀하신대로 insert 이후 id로 다시 조회하여 반환하는 방법은 db 내부에서 값이 변경되어있을 수 있다 등의 이유가 있을 수 있는데요. 밑에 말씀해주신 내용은 과한 가능성인 것 같습니다. 해당 이유로는 db쪽 뿐만아니라 현재 어플리케이션에 있는 모든 코드에 해당할 수 있는 문제인 것 같아요.

jdbcTemplate이 반환된 것에 대해서 테스트하고, 그걸 믿고 사용할 수 있느냐 없느냐의 관점에서만 보시면 좋을 것 같습니다.

RaZel713 replied:
insert 이후 반환된 id의 값을 넣은 객체를 반환할 수 있도록 변경하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/service/RoomThemeService.java:
```diff
@@ -0,0 +1,73 @@
+    public List<PopularThemeResponse> findPopularThemes() {
+        final LocalDate currentDate = LocalDate.now();
+        final LocalDate start = currentDate.minusDays(8);
+        final LocalDate end = currentDate.minusDays(1);
```
8과 1은 의미있는 상수로 변경해야겠네요.

RaZel713 replied:
`PopularThemeSelectionCriteria`를 작성하는 과정에 있어

`private static final int POPULAR_THEME_SELECTION_DURATION = 7;`

위와 같이 상수로 변경하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/service/RoomThemeService.java:
```diff
@@ -0,0 +1,73 @@
+    public List<PopularThemeResponse> findPopularThemes() {
+        final LocalDate currentDate = LocalDate.now();
+        final LocalDate start = currentDate.minusDays(8);
+        final LocalDate end = currentDate.minusDays(1);
```
날짜 범위를 나타내는 도메인이 될 수 있겠네요.

RaZel713 replied:
`PopularThemeSelectionCriteria` 인기 테마 선택 기준이라는 도메인을 작성하여 
날짜의 범위를 관리하게 수정하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -1,36 +1,79 @@
+            if (bookedTime.contains(reservationTime)) {
+                responses.add(AvailableReservationTimeResponse.from(reservationTime, true));
+                continue;
+            }
```
> 한 메서드에 오직 한 단계의 들여쓰기만 한다.

RaZel713 replied:
``` java
// AvailableReservationTime 
public AvailableReservationTime(final List<ReservationTime> bookedTimes, final ReservationTime reservationTime) {
        validate(bookedTimes, reservationTime);
        this.reservationTime = reservationTime;
        this.bookedStatus = isBooked(bookedTimes, reservationTime);
}

private boolean isBooked(final List<ReservationTime> bookedTimes, final ReservationTime reservationTime) {
        if (bookedTimes.isEmpty()) {
            return false;
        }

        return bookedTimes.contains(reservationTime);
}

// ReservationTimeService
return reservationTimes.stream()
                .map(time -> new AvailableReservationTime(bookedTimes, time))
                .toList();
```
이용 가능한 예약 시간을 나타내는 도메인인 `AvailableReservationTime` 을 작성하여 
예약 시간과 예약 상태에 대하여 관리하게끔 수정하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,52 +1,102 @@
+                .orElseThrow(() -> new NotExistedValueException("존재하지 않는 테마 입니다"));
+    }
+
+    private void validatePastDateAndTime(final LocalDate date, final ReservationTime time) {
+        final LocalDate currentDate = LocalDate.now();
+
+        final boolean isPastDate = date.isBefore(currentDate);
+        final boolean isPastTime = date.isEqual(currentDate) && time.getStartAt().isBefore(LocalTime.now());
+
+        if (isPastDate || isPastTime) {
+            throw new PharmaceuticalViolationException("과거 시점은 예약할 수 없습니다");
+        }
```
> 게터/세터/프로퍼티를 쓰지 않는다.

`Reservation`이 처리해야하는 로직이 아닌가요? `Reservation`에서 getter를 가져와서 외부에서 처리하면 `Reservation`은 도메인이 맞나요? Service와 Domain이 처리해야하는 로직의 차이는 무엇인가요?

RaZel713 replied:
``` java
// ReservationService
private void validatePastDateAndTime(final Reservation reservation) {
        if (reservation.validatePastDateAndTime()) {
            throw new PharmaceuticalViolationException("과거 시점은 예약할 수 없습니다");
        }
}

// Reservation
 public boolean validatePastDateAndTime() {
        final LocalDate currentDate = LocalDate.now();

        final boolean isPastDate = date.isBefore(currentDate);
        final boolean isPastTime = date.isEqual(currentDate) && time.validatePastTime();

        return isPastDate || isPastTime;
}
```

ReservationService에서 하던 과거 시점 검증 로직을 Reservation 도메인 내부로 옮겨, 
객체 스스로 자신의 상태를 검증하도록 리팩터링하였습니다.

프로그래밍 과정에서 Reservation 생성자 안에서 과거 시점 검증을 하였으나, 
과거의 기록을 불러오는 과정에 있어 (조회-역직렬화) 오류가 발생하는 것을 확인하여 해당 로직을 service로 옮기는 과정에서
도메인에 로직을 호출하여 처리한다는 생각을 해보지 못하고 서비스에서 처리하게 되었습니다..🥹

좀 더 꼼꼼하게 확인하도록 하겠습니다!

jinyoungchoi95 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,52 +1,102 @@
+        final LocalDate currentDate = LocalDate.now();
+
+        final boolean isPastDate = date.isBefore(currentDate);
+        final boolean isPastTime = date.isEqual(currentDate) && time.getStartAt().isBefore(LocalTime.now());
```
currentDate가 있는데 `LocalDate.now()`를 또 생성하고 있네요.

RaZel713 replied:
https://github.com/woowacourse/spring-roomescape-member/pull/169#discussion_r2070122040
해당 피드백 반영 과정에서 수정하였습니다.

jinyoungchoi95 commented on src/test/java/roomescape/RoomescapeApplicationTest.java:
위에 말씀드렸듯 테스트가 전반적으로 부족합니다. 도메인 테스트도 없네요. 

레벨 1에서의 테스트를 다시 고민해보시면 좋겠습니다.

RaZel713 replied:
도메인 테스트와 DAO에 대한 테스트를 추가 작성하였습니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 라젤 😄

방탈출 사용자 예약 미션의 리뷰어를 맡은 오리입니다.

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

RaZel713 committed:
refactor: 상수의 접근 제한자를 private 으로 변경

RaZel713 committed:
refactor: 매직 넘버 상수 처리

RaZel713 committed:
feat: 인기 테마 날짜 범위 도메인 추가

- 인기 테마 선정 기준일과 선정 기간 일 수를 받아 인기 테마 선정 날짜 범위를 계산하는 도메인 클래스를 추가한다.

RaZel713 committed:
refactor: 과거 시점 검사 로직을 Reservation 도메인으로 이동

- 서비스에서 처리하던 예약 시점 유효성 검사를 도메인 내부 메서드로 이동하여 도메인이 자신의 상태를 스스로 판단할 수 있도록 한다.

RaZel713 committed:
refactor: 예약 가능 시간 응답 생성 로직을 변수 추출하여 가독성 향상

RaZel713 committed:
refactor: 예약 가능 시간 응답 생성 메서드 네이밍 구체화

- AvailableReservationTimeResponse의 정적 팩토리 메서드명을 from → fromReservationTimeWithBookingStatus로 변경한다.
- 메서드의 인자가 예약 시간과 예약 여부를 포함하므로, 의미를 명확히 드러내도록 개선한다.

RaZel713 committed:
refactor: 인기 테마 개수 제한을 쿼리에서 제거하고 서비스 로직으로 이동

- 쿼리에 비즈니스 요구사항이 섞여 있으면 요구사항 변경 시 쿼리 수정이 필요해 유지 보수가 어렵다.
- 따라서 쿼리에서 LIMIT 10을 제거하고, Java 코드에서 상위 10개만 추려내도록 변경한다.

RaZel713 committed:
refactor: 예약 도메인 정렬 순서 변경

RaZel713 committed:
test: 예약 도메인 테스트 추가

RaZel713 committed:
refactor: 예약 시간 도메인 정렬 순서 변경

RaZel713 committed:
test: 예약 시간 도메인 테스트 추가

RaZel713 committed:
test: 방 테마 도메인 테스트 추가

RaZel713 committed:
test: 인기 테마 선택 기준 도메인 테스트 추가

RaZel713 committed:
refactor: 서비스가 도메인을 반환하고 컨트롤러에서 DTO로 변환하도록 수정

- 서비스 계층에서 컨트롤러의 Response DTO에 대한 의존 제거한다
- 컨트롤러에서 도메인을 받아 DTO로 변환하는 책임을 수행한다
- 계층 간 의존 방향을 컨트롤러 → 서비스 라는 것을 명확히 한다

RaZel713 committed:
test: 기능별 http 작성 및 패키지 이동

RaZel713 committed:
test: 예약 DAO 테스트 추가

RaZel713 committed:
test: 예약 시간 DAO 테스트 추가

RaZel713 committed:
test: 방 테마 DAO 테스트 추가

RaZel713 committed:
refactor: 클래스 네이밍 수정

RaZel713 committed:
test: 변수명 변경

RaZel713 committed:
feat: 사용 중인 시간을 삭제할 경우 발생하는 예외 추가 및 비즈니스 규칙 위반 예외 네이밍 변경

RaZel713 committed:
refactor: '@ResponseStatus' 애너테이션을 사용하여 메서드 별 HTTP 상태 코드 명시

RaZel713 committed:
refactor: 줄바꿈 변경

RaZel713 committed:
refactor: 유효성 검사 메서드 추출로 메서드 간 추상화 수준 일치

RaZel713 committed:
refactor: 메서드 네이밍 변경

RaZel713 reviewed (commented):
안녕하세요 ~ 오리!
꼼꼼하게 피드백 주셔서 감사해요! 😊
피드백 주신 부분들에 대하여 반영하고 comment 달았어요!

최대한 핵심만 요약해서 작성해봤는데, 부족한 부분 있다고 생각하시면 피드백 주시면 좋을 것 같아요!

리뷰 반영에 따른 변경 사항 커밋 범위 : [step1-3 1차 피드백](https://github.com/woowacourse/spring-roomescape-member/pull/169/files/354bc89fc7f9567387a4453cd5e59ef0f56e3847..29c27d4c98380c05770756dc53b2a3847b413383)

jinyoungchoi95 commented on src/main/java/roomescape/controller/api/ReservationController.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @GetMapping
+    @ResponseStatus(HttpStatus.OK)
```
ok를 명시하지 않아도 spring에선 기본으로 200을 내려주니 명시할 필요가 있을까요?

RaZel713 replied:
코드 스타일의 일관성과 의도 전달의 명확성을 위하여 사용하였습니다!

jinyoungchoi95 replied:
의도 전달의 명확성이라기엔 spring docs에 나온 스펙대로라면 없이 반환하는 것이 이미 명확하게 200인데 왜 추가했는지에 대한 의문을 줄 수 있겠네요.

jinyoungchoi95 commented on src/main/java/roomescape/controller/dto/response/RoomThemeResponse.java:
```diff
@@ -0,0 +1,10 @@
+
+public record RoomThemeResponse(long id, String name, String description, String thumbnail) {
+
+    public static RoomThemeResponse from(final RoomTheme theme) {
```
final이 붙은곳과 안붙은곳의 차이가 있을까요?

RaZel713 replied:
`record`의 필드는 이미 불변이므로, `final`을 붙이지 않아도 됩니다.
하지만 DTO의 정적 팩토리 메서드의 파라미터는 불변이 아니므로, 
해당 값을 변경할 수 없다는 의미를 드러내기 위하여 final 키워드를 활용하여 표현하였습니다.

- DTO는 데이터를 외부로 전달하거나 계층 간 이동시키기 위한 객체로, 상태 변경이 아닌 데이터 전달 자체에 목적이 있습니다.
- DTO를 생성할 때 사용하는 파라미터는 해당 객체를 구성하는 값을 전달하는 용도이므로, 메서드 내부에서 값이 변경되지 않도록 final 키워드를 명시하는 것이 명확성과 실수 방지 측면에서 좋다고 판단하여 사용하였습니다.

<details>
<summary> 사용 위치에 따른 final 키워드의 의미 </summary>

- **변수**:  한번만 초기화 가능하고, 그 이후 변경이 불가능하다.
- **메서드**: 오버라이딩을 금지한다. 하위 클래스에서 재정의가 불가능하다.
- **클래스**: 상속을 금지한다. 다른 클래스가 해당 클래스를 확장할 수 없다.
- **매개변수**: 메서드 안에서 값을 변경할 수 없다.

</details>

(전체적으로 final 키워드가 빠진 부분들을 리팩터링하였습니다.)

추신) 질문한 의도에 맞는 답변일까요?

jinyoungchoi95 replied:
```java
    public List<AvailableReservationTimeResponse> findAvailableTime(@RequestParam(value = "date") LocalDate date,
```
final 이 안달린 파라메터들이 있어 해당 부분에 대해 말씀드린 것이었어요.

jinyoungchoi95 commented on src/main/java/roomescape/domain/PopularThemeSelectionCriteria.java:
```diff
@@ -0,0 +1,28 @@
+    private final LocalDate startDay;
+    private final LocalDate endDay;
+
+    public PopularThemeSelectionCriteria(final LocalDate startBaseDate, final int durationInDays) {
```
`durationInDays`가 음수가 들어오면 어떻게 되나요?

RaZel713 replied:
startBaseDate, durationInDays에 대한 값 검증을 작성하지 않고 넘어간 것 같습니다 🥹

startBaseDate, durationInDays 모두 필드의 값 계산에 있어 유효한 값이 들어오지 않으면 생성에 문제가 발생하므로, 
유효성 검사 메서드를 작성하여 생성자 내부에서 호출할 수 있게 수정하였습니다.
``` java
private void validateInvalidInput(final LocalDate startBaseDate, final int durationInDays) {
        if (startBaseDate == null) {
            throw new InvalidInputException("시작 기준일은 null이 될 수 없습니다");
        }
        if (durationInDays <= 0) {
            throw new InvalidInputException("인기 테마 선정 기준 기한 일수는 음수나 0이 될 수 없습니다");
        }
}

```

jinyoungchoi95 commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,27 @@
+CREATE TABLE theme
+(
+    id          BIGINT       NOT NULL AUTO_INCREMENT,
+    theme_name  VARCHAR(255) NOT NULL,
+    description VARCHAR(255) NOT NULL,
+    thumbnail   VARCHAR(255) NOT NULL,
```
db에는 글자수 제한이 있는데 도메인에는 없네요.

RaZel713 replied:
글자 수 제한을 추가하여, 제한 길이를 벗어나는 값이 들어오면 `InvalidInputException`이 발생할 수 있게 처리하였습니다.


```java
private void validateValidLength(final String name, final String description, final String thumbnail) {
        if (name.length() > MAX_LENGTH) {
            throw new InvalidInputException("테마 명은 255자를 초과할 수 없습니다.");
        }
        if (description.length() > MAX_LENGTH) {
            throw new InvalidInputException("테마 상세 설명은 255자를 초과할 수 없습니다.");
        }
        if (thumbnail.length() > MAX_LENGTH) {
            throw new InvalidInputException("썸네일 주소는 255자를 초과할 수 없습니다.");
        }
}
```

데이터 베이스와 도메인/비즈니스의 제한을 맞출 생각을 해보지 못한 것 같습니다.
알려주셔서 감사합니다!😊

jinyoungchoi95 commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,42 @@
+import roomescape.exception.custom.NotExistedValueException;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
`GlobalExceptionHandler`에서 핸들링되지 않는 예외는 응답이 어떻게 내려갈까요?

RaZel713 replied:
`GlobalExceptionHandler`에서 핸들링 되지 않는 예외는 스프링에서 자체적으로 처리 하여, 

• HTTP 상태 코드: `500 Internal Server Error`
• 응답 Body 
```java
{
  "timestamp": "2025-05-07T12:34:56.789+00:00",
  "status": 500,
  "error": "Internal Server Error",
  "path": "/api/some-path"
}
```

위와 같은 응답이 내려가는 것을 알 수 있었습니다!
`500 Internal Server Error` 상태 코드는 
서버에서 요청을 처리하는 도중 예상하지 못한 에러가 발생했음을 나타내므로,
미처 정의되지 않는 예외를 표현하는데 적합하다고 생각합니다.

```java
@ExceptionHandler
    public ResponseEntity<String> handleException(Exception e) {
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(ERROR_PREFIX + e.getMessage());
}
```
따라서 위의 코드를 추가하여 정의한 예외를 제외한 예외를 처리하도록 하였습니다!
(해당 코드가 없어도 자동으로 처리되지만, 에러 메시지 형식의 일관성을 위해 명시적으로 작성하였습니다.)

jinyoungchoi95 commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    private void validatePastDateAndTime(final Reservation reservation) {
+        if (reservation.validatePastDateAndTime()) {
```
도메인에서 바로 예외를 발생시키지 않고 서비스에서 예외를 던지게 하는 이유가 있을까요?

RaZel713 replied:
### **예외를 도메인에서 바로 던지지 않고 판단만 하도록 분리한 이유**

- 시점 판별이라는 기능이 **예약 생성 외에도 다양한 곳에서 활용**될 수 있기 때문입니다. 
   (예를 들어, 사용자가 예약 내역 중 과거 예약만 필터링해서 조회하는 기능을 만들거나, 
   지난 예약만 통계로 집계하고자 할 때에도 이 메서드를 재사용할 수 있습니다.)

- 예약의 날짜와 시간 값이 과거 시점일 경우 문제가 되는 상황은 
예약 생성 시에만 유효성 문제가 발생하므로, 조회나 삭제 같은 다른 기능에서는 과거 시점의 예약은 문제가 되지 않습니다. 

- 만약 도메인 내부에서 과거 시점이면 예외를 던지도록 해버리면, 해당 도메인을 사용하는 모든 기능에서 과거 시점의 예약이 문제가 되는 것처럼 느껴질 수 있다고 생각합니다. **예외를 발생시킬지 여부는 사용 맥락에 따라 달라지므로** 이를 도메인에 강제하지 않고 서비스 계층에서 처리하는 것이 더 적합한 설계라고 생각합니다.

---

이로 인해 도메인에서는 단순히 해당 예약이 과거 시점인지 아닌지를 판단하는 메서드를 만들어 boolean 값을 반환하게 하고, 예외 처리는 서비스 계층에서 담당하도록 설계하였습니다.

jinyoungchoi95 replied:
언젠가 재사용될 것을 기대하고 로직을 분리하지는 않으셨으면 좋겠습니다.

`Reservation`을 하나의 객체가 아닌 최초 생성 예약과 생성된 예약으로 구현체를 나누어보면 되지 않을까요? service에 domain 로직이 담기진 않았는지, 레벨 1이었다면 어떻게 했을 것인지 계속 고민해보시면 좋겠어요.

jinyoungchoi95 commented on src/main/java/roomescape/service/RoomThemeService.java:
```diff
@@ -0,0 +1,77 @@
+        PopularThemeSelectionCriteria criteria =
+                new PopularThemeSelectionCriteria(LocalDate.now(), POPULAR_THEME_SELECTION_DURATION);
+
+        List<RoomTheme> popularThemes = themeDAO.findPopularThemes(criteria.getStartDay(), criteria.getEndDay());
+
+        return popularThemes.subList(0, Math.min(POPULAR_THEMES_TOP_COUNT, popularThemes.size()));
```
top count를 dao에 요청해서 가져오면 쿼리 수도 줄어드는 것 아닐까요?

RaZel713 replied:
제한 숫자(LIMIT)를 파라미터로 받는 방식으로 구현하면,
비즈니스 요구사항이 변경되더라도 쿼리의 변경 없이 수정할 수 있어 유연성이 높아지고,
필요한 데이터만 조회하므로 성능 측면에서도 훨씬 좋다고 생각합니다!

초기에 피드백을 받았을 때 service에서 제한에 맞춰 subList로 잘라 반환하는 방법밖에 고려하지 않은 것 같습니다.

다음부터 더 나은 방법이 있는지 고민해보고 적용해보도록 하겠습니다!
알려주셔서 감사합니다! ☺️ 

``` java
public List<RoomTheme> findPopularThemes(final LocalDate start, final LocalDate end, final int limit) {

        final String query = """
                        SELECT
                            t.id,
                            t.theme_name,
                            t.description,
                            t.thumbnail
                        FROM theme AS t
                        JOIN reservation AS r
                        ON r.theme_id = t.id
                        WHERE r.date BETWEEN ? AND ?
                        GROUP BY t.id
                        ORDER BY COUNT(r.id) DESC
                        LIMIT ?
                """;
        return jdbcTemplate.query(query, THEME_ROW_MAPPER, start, end, limit);
}
```

jinyoungchoi95 commented on src/test/java/roomescape/dao/JdbcReservationDaoTest.java:
```diff
@@ -0,0 +1,116 @@
+        List<Reservation> result = reservationDAO.findAll();
+
+        // then
+        assertThat(result).hasSize(1);
+        assertThat(result.getFirst().getId()).isEqualTo(1L);
+        assertThat(result.getFirst().getDate()).isEqualTo(LocalDate.of(2025, 4, 28));
+        assertThat(result.getFirst().getTime()).isEqualTo(new ReservationTime(1L, LocalTime.of(10, 0)));
```
`assertAll`이어야겠네요.

RaZel713 replied:
피드백 주신 내용 반영하여, 
`assertAll`을 사용해 여러 검증을 한 번에 확인할 수 있도록 테스트를 수정하였습니다.

jinyoungchoi95 commented on src/test/java/roomescape/domain/ReservationTimeTest.java:
```diff
@@ -0,0 +1,43 @@
+
+    @DisplayName("해당 예약 시간이 현재 시간보다 과거인 경우 true를 반환한다")
+    @Test
+    void validatePastTimeTest1() {
+        // given
+        LocalTime pastTime = LocalTime.now().minusMinutes(1);
+        ReservationTime reservationTime = new ReservationTime(pastTime);
+
+        // when // then
+        assertThat(reservationTime.validatePastTime()).isTrue();
+    }
+
+    @DisplayName("해당 예약 시간이 현재 시간보다 미래인 경우 false를 반환한다")
+    @Test
+    void validatePastTimeTest2() {
```
test1, test2일것이라면 영문 테스트 메서드와 `@DisplayName`를 모두 사용할 필요가 있을까요?

RaZel713 replied:
테스트 메서드 명을 테스트에 맞게 수정하였습니다!

jinyoungchoi95 commented on src/main/java/roomescape/controller/api/ReservationTimeController.java:
`Controller`테스트가 없는 이유가 있을까요?

RaZel713 replied:
Controller 테스트를 추가하였습니다.

jinyoungchoi95 reviewed (commented):
안녕하세요 라젤 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

jinyoungchoi95 reviewed (changes requested):
안녕하세요 라젤 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

RaZel713 committed:
refactor: 도메인의 생성자 파라미터 유효성 검증 메서드 네이밍 변경

RaZel713 committed:
feat: 인기 테마 선정 기준 도메인의 생성자 파라미터 유효성 검증 메서드 추가

RaZel713 committed:
refactor: final 키워드 추가

RaZel713 committed:
feat: 예약자 명 길이 검증 메서드 추가

RaZel713 committed:
feat: 테마 명, 테마 설명, 썸네일 주소 길이 검증 메서드 추가

RaZel713 committed:
test: assertAll 적용

RaZel713 committed:
test: 데이터 베이스 연결 테스트 추가

RaZel713 committed:
test: 어드민 페이지 컨트롤러 테스트 추가

RaZel713 committed:
test: 예약 페이지 컨트롤러 테스트 추가

RaZel713 committed:
test: 예약 시간 테스트 데이터 변경

RaZel713 committed:
test: 예약 테마, 예약 테스트 데이터 변경

RaZel713 committed:
test: 예약 컨트롤러 테스트 추가

RaZel713 committed:
test: 예약 시간 컨트롤러 테스트 추가

RaZel713 committed:
test: 방 테마 컨트롤러 테스트 추가

RaZel713 committed:
feat: 정의된 예외를 제외한 모든 오류 발생 시 500 Internal Server Error 상태 코드 반환

- 비즈니스 로직에서 정의한 예외를 제외한 예기치 못한 오류 발생 시 500 상태 코드 반환한다

RaZel713 committed:
refactor: 도메인 패키지 분리

- 예약, 예약 시간, 테마 관련 클래스를 도메인 패키지 하위에 분리한다

RaZel713 committed:
test: 테스트 데이터 변경에 따른 HTTP 파일 수정

RaZel713 committed:
refactor: 인기 테마 조회 시 상위 개수 제한을 파라미터로 받도록 변경

RaZel713 committed:
test: 테스트 메서드 명 변경

RaZel713 committed:
refactor: 컨트롤러와 서비스 DTO 네이밍 변경

RaZel713 committed:
refactor: 예약, 예약 시간 도메인의 과거 시점 판단 메서드 명 변경

- 시점을 판별하여 boolean 값을 반환하므로 is로 시작하는 메서드 명으로 변경한다

RaZel713 committed:
refactor: 서비스 DTO 네이밍 변경

RaZel713 committed:
refactor: service에서 insert 이후 반환된 id의 값을 넣은 객체를 반환

RaZel713 committed:
refactor: 컨트롤러를 rest와 view 패키지로 분리

RaZel713 reviewed (commented):
안녕하세요 오리!
step1-3 두번째 피드백 내용 모두 반영했습니다.
부족한 부분이나 제가 잘못 이해한 부분 있으면 알려주시면 좋을 것 같아요!😊

jinyoungchoi95 commented on src/main/java/roomescape/controller/rest/ReservationController.java:
```diff
@@ -0,0 +1,51 @@
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public ReservationResponse addReservation(
+            @RequestBody roomescape.controller.dto.request.CreateReservationRequest request) {
```
```suggestion
            @RequestBody CreateReservationRequest request) {
```

jinyoungchoi95 commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,47 @@
+
+    @ExceptionHandler
+    public ResponseEntity<String> handleException(Exception e) {
+        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(ERROR_PREFIX + e.getMessage());
```
`ResponseEntity`에서 제공하는 기본 빌더 메서드를 사용하지 않는 이유가 있을까요?

jinyoungchoi95 commented on src/test/java/roomescape/service/ReservationTimeServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+class ReservationTimeServiceTest {
+
+    @Autowired
+    ReservationTimeService reservationTimeService;
```
```suggestion
    private ReservationTimeService reservationTimeService;
```

jinyoungchoi95 commented on src/test/java/roomescape/service/ReservationTimeServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+    @Test
+    @DisplayName("시간 데이터를 추가할 수 있어야 한다")
+    void addReservationTimeTest() {
+        // given // when
+        final CreateReservationTimeServiceRequest creation = new CreateReservationTimeServiceRequest(
+                LocalTime.of(13, 0));
+        reservationTimeService.addReservationTime(creation);
```
given과 when을 분리해야겠네요

jinyoungchoi95 commented on src/test/java/roomescape/service/ReservationTimeServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+    @Test
+    @DisplayName("시간 데이터를 조회할 수 있어야 한다")
+    void findAllReservationTimesTest() {
+        // given // when
+        final List<ReservationTime> result = reservationTimeService.findAllReservationTimes();
```
given이 없으면 주석에 추가하지 않아도 됩니다. bdd는 주석을 강제하는 패턴이 아니에요.

jinyoungchoi95 reviewed (approved):
안녕하세요 라젤 😄

몇가지 코멘트 남겨두었으니 다음 미션 진행하면서 함께 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
